code/modules/hydroponics/compost_bin.dm

```diff
@@ -0,0 +1,194 @@
+	if(user.a_intent == INTENT_HARM)
+		return ..()
+
+	if(istype(O, /obj/item/storage/bag/plants))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more biomass!</span>")
+			return
+
+		var/obj/item/storage/bag/plants/PB = O
+		for(var/obj/item/reagent_containers/food/snacks/grown/G in PB.contents)
+
+			if(biomass >= biomass_capacity)
+				break
+			PB.remove_from_storage(G, src)
+			make_biomass(G)
+
+		// start composting after plants are inserted
+		compost()
+
+		if(biomass < biomass_capacity)
+			to_chat(user, "<span class='info'>You empty [PB] into [src].</span>")
+		else
+			to_chat(user, "<span class='info'>You fill [src] to its capacity.</span>")
+
+		SStgui.update_uis(src)
+		update_icon_state()
+		return TRUE
+
+	if(istype(O, /obj/item/reagent_containers/food/snacks/grown))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more plants!</span>")
+			return
+		if(!user.unEquip(O))
+			return
+
+		O.forceMove(src)
+		make_biomass(O)
+		// start composting after plants are inserted
+		compost()
+		to_chat(user, "<span class='info'>You put [O] in [src].</span>")
+		SStgui.update_uis(src)
+		update_icon_state()
+		return TRUE
+
+	to_chat(user, "<span class='warning'>You cannot put this in [name]!</span>")
+
+/obj/machinery/compost_bin/attack_hand(mob/user)
+	ui_interact(user)
+
+
+/obj/machinery/compost_bin/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "CompostBin", "CompostBin", 390, 300, master_ui, state)
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+
+/obj/machinery/compost_bin/ui_data(mob/user)
+	var/list/data = list()
+	data["biomass"] = biomass
+	data["biomass_capacity"] = biomass_capacity
+	data["compost"] = compost
+	data["compost_capacity"] = compost_capacity
+	return data
+
+// Start composting if there is enough biomass and space for compost
+/obj/machinery/compost_bin/proc/compost()
+	// Prevents the compost bin from starting to compost again while already composting
+	if(composting)
+		return
+	if(compost >= compost_capacity || biomass <= 0)
+		return
+	composting = TRUE
+	addtimer(CALLBACK(src, PROC_REF(convert_biomass)), 10 SECONDS)
+
+// Convert biomass to compost, then continue composting
+/obj/machinery/compost_bin/proc/convert_biomass()
+	//converts 20% of the biomass to compost each cycle, unless there isn't enough comopst space or there is 10 or less biomass
+	var/conversion_amount = clamp(DECAY * biomass, min(MIN_CONVERSION, biomass), compost_capacity - compost)
+	biomass -= conversion_amount
+	compost += conversion_amount
+	update_icon_state()
+	SStgui.update_uis(src)
+	if(compost >= compost_capacity || biomass <= 0)
+		composting = FALSE
+	else
+		addtimer(CALLBACK(src, PROC_REF(convert_biomass)), 10 SECONDS)
```

Comment:
wouldn't this cause more processing? As is the compost bin only checks for compost conditions when filled or emptied, and loops until it can no longer compost. It seems that by using process() I'd be adding extra condition checks every tick, no?

Reply:
Timers are actually pretty expensive, processing is not. I think its better in this case to use processing

Reply:
Righto, will do.